[ML] Transforms: Refactor validators and add unit tests.

walterra opened the pull request:
## Summary

Part of #173301.

Splits up form validators and adds more unit tests.

### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

walterra commented:
buildkite build this

walterra commented:
buildkite build this

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

darnautov commented on x-pack/plugins/transform/public/app/common/validators/frequency_validator.test.ts:
```diff
@@ -0,0 +1,47 @@
+import { frequencyValidator } from './frequency_validator';
+
+describe('Transform: frequencyValidator()', () => {
+  const transformFrequencyValidator = (arg: string) => frequencyValidator(arg).length === 0;
```
Not sure if we should define a function inside of a test suite and use it for assertion 🤔 
Usually, you test a function that is used in runtime.

walterra replied:
You're right, I removed it in here [0e8ab71](https://github.com/elastic/kibana/pull/173736/commits/0e8ab71711b11ae83d836066592f51e22bfd2216), it's now in line with the other test files.

darnautov commented on x-pack/plugins/transform/public/app/common/validators/integer_range_10_to_10000_validator.ts:
```diff
@@ -0,0 +1,24 @@
+  }
+);
+
+export const integerRange10To10000Validator: Validator = (value) =>
```
This function seems too specific to me. Should we create a generic validator for string representation of numbers? Otherwise, we end up with loads of similar functions for each specific case.

walterra replied:
Looks like this was no longer in use anyway, so I removed it in [90e20ac](https://github.com/elastic/kibana/pull/173736/commits/90e20ac7f6038e4243a0645384e1ee88ffecbb8f).

darnautov commented on x-pack/plugins/transform/public/app/common/validators/transform_settings_number_of_retries_validator.ts:
```diff
@@ -0,0 +1,24 @@
+);
+
+export const transformSettingsNumberOfRetriesValidator: Validator = (value) =>
+  !(value + '').includes('.') &&
```
Can this be a part of a generic validator?

darnautov commented on x-pack/plugins/transform/public/app/common/validators/transform_settings_page_search_size_validator.ts:
```diff
@@ -0,0 +1,29 @@
+ * @param value User input value.
+ */
+export const transformSettingsPageSearchSizeValidator: Validator = (value) =>
+  !(value + '').includes('.') &&
```
Can this be a part of a generic validator?

darnautov commented on x-pack/plugins/transform/public/app/common/validators/string_validator.ts:
```diff
@@ -0,0 +1,22 @@
+
+import type { Validator } from './types';
+
+export const stringValidator: Validator = (value, isOptional = true) => {
```
we already have a generic [requiredValidator](https://github.com/elastic/kibana/blob/8453fe820a7bd0b264b3963f547a5092b5e811d5/x-pack/plugins/ml/common/util/validators.ts#L61)

darnautov commented on x-pack/plugins/transform/public/app/common/validators/retention_policy_max_age_validator.test.ts:
```diff
@@ -0,0 +1,45 @@
+import { retentionPolicyMaxAgeValidator } from './retention_policy_max_age_validator';
+
+describe('Transform: retentionPolicyMaxAgeValidator()', () => {
+  const transformRetentionPolicyMaxAgeValidator = (arg: string) =>
+    retentionPolicyMaxAgeValidator(arg).length === 0;
```
I reckon we should not write unit tests for functions created in the test suites itself

walterra replied:
Fixed in [0193145](https://github.com/elastic/kibana/pull/173736/commits/01931450a6b508122f7138118dbd61a556d22528).

darnautov commented on x-pack/plugins/transform/public/app/common/validators/is_valid_frequency.ts:
```diff
@@ -0,0 +1,29 @@
+import type { ParsedDuration } from './types';
+
+// only valid if value is up to 1 hour
+export function isValidFrequency(arg: unknown): arg is ParsedDuration {
```
It would be useful to have a generic duration validator, e.g.
```
const getDurationValidator = ({  minDurationInMs: number; maxDurationInMs: number; allowedUnits: stiring[] })
```

darnautov commented on x-pack/plugins/transform/public/app/common/validators/is_retention_policy_max_age.ts:
```diff
@@ -0,0 +1,36 @@
+
+  // only valid if value is equal or more than 60s
+  // supported time units: https://www.elastic.co/guide/en/elasticsearch/reference/master/common-options.html#time-units
+  return (
+    (timeUnit === 'nanos' && number >= RETENTION_POLICY_MIN_AGE_SECONDS * 1000000000) ||
+    (timeUnit === 'micros' && number >= RETENTION_POLICY_MIN_AGE_SECONDS * 1000000) ||
+    (timeUnit === 'ms' && number >= RETENTION_POLICY_MIN_AGE_SECONDS * 1000) ||
+    (timeUnit === 's' && number >= RETENTION_POLICY_MIN_AGE_SECONDS) ||
+    ((timeUnit === 'm' || timeUnit === 'h' || timeUnit === 'd') && number >= 1)
+  );
```
that's where a generic duration validator would come in handy

darnautov commented on x-pack/plugins/transform/public/app/common/validators/is_json_string.ts:
```diff
@@ -0,0 +1,23 @@
+  try {
+    return !!JSON.parse(value);
+  } catch (e) {
+    // eslint-disable-next-line no-console
+    console.error(`JSON is invalid.\n${e}`);
```
do we need to log it?

walterra replied:
removed it in [eb68c06](https://github.com/elastic/kibana/pull/173736/commits/eb68c06de6aba3d2ab537b1a9538b7faf8db33e6).

darnautov commented on x-pack/plugins/transform/public/app/common/validators/is_json_string.ts:
```diff
@@ -0,0 +1,23 @@
+  if (typeof value !== 'string') return false;
+
+  try {
+    return !!JSON.parse(value);
```
`JSON.parse(true)` returns `true` as well. So you should also check for type `string`.

walterra replied:
We already have a check in there to only allows string. I added another jest test to check for `true` here [5321409](https://github.com/elastic/kibana/pull/173736/commits/532140900e590d0163e1be5704c4bddb62d8f363).

darnautov commented on x-pack/plugins/transform/public/app/common/validators/is_continuous_mode_delay.ts:
```diff
@@ -0,0 +1,15 @@
+ * Doesn't allow floating intervals.
+ * @param value User input value.
+ */
+export function isContinuousModeDelay(value: string): boolean {
```
we have a generic [patternValidator](https://github.com/elastic/kibana/blob/8453fe820a7bd0b264b3963f547a5092b5e811d5/x-pack/plugins/ml/common/util/validators.ts#L33)

walterra replied:
Looks like `patternValidator` lives in the `ml` plugin, not `transform`. We could at one point create a package for validators and share that among plugins, but I'm afraid it's not something I'm going to pick up in this PR 😅 .

darnautov commented on x-pack/plugins/transform/public/app/common/validators/integer_range_minus_1_to_100_validator.ts:
```diff
@@ -0,0 +1,24 @@
+  }
+);
+
+export const integerRangeMinus1To100Validator: Validator = (value) =>
+  !(value + '').includes('.') &&
+  numberValidator({ min: -1, max: 100, integerOnly: true })(+value) === null
+    ? []
+    : [numberRangeMinus1To100NotValidErrorMessage];
```
can we use a generic validator I mentioned before? 
another important moment to consider, depending on the locale, a comma might be used as a decimal separator

darnautov commented on x-pack/plugins/transform/public/app/common/validators/integer_above_zero_validator.ts:
```diff
@@ -0,0 +1,23 @@
+);
+
+export const integerAboveZeroValidator: Validator = (value) =>
+  !(value + '').includes('.') && numberValidator({ min: 1, integerOnly: true })(+value) === null
```
Can we use a generic validator and avoid creating a function for each specific case?

walterra replied:
I'm not sure I fully get the suggestion. This is making use of the generic validator `numberValidator`, it then sets this up as a wrapper for a specific use case required by the edit transform flyout to satisfy the spec of the request object to update a transform. Also note this isn't new code, all the validators have just been moved to individual files and a corresponding jest test file, the intent of this PR wasn't to change any logic but just to spread out the existing validators into separate files. Can you share more details about how you envision that generic validator?

darnautov replied:
Why do you need the `!(value + '').includes('.')` part? The generic validator already checks for integers with `ingegerOnly`.

walterra replied:
We have a test assertion in there that would fail if remove that part: `expect(integerAboveZeroValidator('1.')).toEqual(['Value needs to be an integer above zero.']);`. I guess it's because `1.` technically passes `Number.isInteger` but we didn't want users to be able to enter `1.`.

darnautov commented on x-pack/plugins/transform/public/app/common/validators/integer_range_minus_1_to_100_validator.ts:
```diff
@@ -0,0 +1,24 @@
+  }
+);
+
+export const integerRangeMinus1To100Validator: Validator = (value) =>
```
Can we use a generic validator and avoid creating a function for each specific case?

walterra commented:
@darnautov I addressed your comments and adjusted the test assertions. As mentioned in one of the comments, I'd like to avoid a larger refactor in this PR, the aim was just to split up larger files into files with 1 validator and corresponding test file. Please have another look.

walterra commented:
@darnautov As discussed, I removed the checks for `!(value + '').includes('.')` in [49f1284](https://github.com/elastic/kibana/pull/173736/commits/49f128477097dd92a60bd6878b6c6dc1e53b60eb)

alvarezmelissa87 reviewed (approved):
LGTM ⚡
